teuthology/orchestra/run.py

```diff
@@ -221,6 +233,73 @@ def __repr__(self):
             name=self.hostname,
             )
 
+class ErrorScanner(object):
+    """
+    Scan for unit tests errors in teuthology.log 
+    """
+    __flag__ = 0
+    def __init__(self):
+        self.ERROR_PATTERNS = {
+            "prev_detected": [re.compile(r"UnitTestError")],
+            "nose": [
+                re.compile(r"ERROR:\s"),
+                re.compile(r"FAIL:\s"),
+            ],
+            "gtest":  [re.compile(r"\[\s\sFAILED\s\s\]")],
+        }
+    
+    def _search_error(self, line, test):
+        test_patterns = self.ERROR_PATTERNS[test]
+
+        for pattern in test_patterns:
+            error = pattern.search(line)
+            if error:
+                is_old_error = self._is_prev_detected_error(line)
+                if not is_old_error:
+                    return line[error.start():].strip()
+        return None
+
+    def _is_prev_detected_error(self, line) -> bool:
+        for detected_pattern in self.ERROR_PATTERNS['prev_detected']:
+            if detected_pattern.search(line):
+                return True
+        return False
+
+    def scan(self, test_names=None):
+        logfile = self._logfile
+        if not logfile or not test_names:
+            return None, None
+
+        error_test = None
+        error_msg = None
+
+        log.debug("__flag__ pointer at: {}".format(ErrorScanner.__flag__))
+        f = open(logfile, 'r')
```

Comment:
Instead of calling `open()` here and `close()` separately later on, we should use a `with` statement. A short explanation for that is with this current implementation, any exception encountered between `open()` and `close()` will mean that `close()` never gets called. If that's confusing or if you're curious to learn more, check out this [very in-depth writeup](https://realpython.com/python-with-statement/#managing-resources-in-python)

Reply:
I see, thank you for sharing this! :)